lighthouse-core/scripts/generate-bundlephobia-database.js

```diff
@@ -0,0 +1,145 @@
+
+const fs = require('fs');
+const getPackageVersionList = require('bundle-phobia-cli').fetchPackageStats.getPackageVersionList;
+const fetchPackageStats = require('bundle-phobia-cli').fetchPackageStats.fetchPackageStats;
+
+const databasePath = '../audits/byte-efficiency/bundlephobia-database.json';
+
+/** @type {Record<string, string[]>} */
+const suggestionsJSON = require('../audits/byte-efficiency/library-suggestions.js').suggestions;
+/** @type string[] */
+/* eslint-disable-next-line max-len */
+const librarySuggestions = Object.keys(suggestionsJSON).map(s => suggestionsJSON[s].concat(s)).flat();
+
+/** @type {Record<string, Record<'lastScraped', number|string> | Record<'repository', string> | Record<string, BundlePhobiaLibrary>>} */
+let database = {};
+if (fs.existsSync(databasePath)) {
+  database = require(databasePath);
+}
+
+/**
+ * Returns true if this library has been scraped from BundlePhobia in the past week.
+ * @param {string} library
+ * @return {boolean}
+ */
+function hasBeenRecentlyScraped(library) {
+  if (!database[library] || database[library].lastScraped === 'Error') return false;
+  return (Date.now() - database[library].lastScraped) / (1000 * 60 * 60 * 24.0) < 7;
+}
+
+/**
+ * Returns true if the object represents valid BundlePhobia JSON.
+ * The version string must not match this false-positive expression: '{number} packages'.
+ * @param {any} library
+ * @return {boolean}
+ */
+function validateLibraryObject(library) {
+  return library.hasOwnProperty('name') &&
+    library.hasOwnProperty('size') &&
+    library.hasOwnProperty('gzip') &&
+    library.hasOwnProperty('description') &&
+    library.hasOwnProperty('repository') &&
+    library.hasOwnProperty('version') &&
+    !library.version.match(/^([0-9]+) packages$/);
+}
+
+/**
+ * Save BundlePhobia stats for a given npm library to the database.
+ * @param {string} library
+ * @param {number} index
+ */
+async function collectLibraryStats(library, index) {
+  return new Promise(async (resolve, reject) => {
+    console.log(`\n◉ (${index}/${librarySuggestions.length}) ${library} `);
+
+    if (hasBeenRecentlyScraped(library)) {
+      console.log(`   ❕ Skipping`);
+      resolve();
+      return;
+    }
+
+    /** @type {Array<BundlePhobiaLibrary>} */
+    const libraries = [];
+    /** @type {string|number} */
+    let lastScraped = Date.now();
+
+    const versions = await getPackageVersionList(library, 10);
+    for (const version of versions) {
+      try {
+        const libraryJSON = await fetchPackageStats(version);
+        if (validateLibraryObject(libraryJSON)) libraries.push(libraryJSON);
+      } catch (e) {
+        console.log(`   ❌ Failed to fetch stats | ${version}`);
+        lastScraped = 'Error';
+      }
+    }
+
+    for (let index = 0; index < libraries.length; index++) {
+      const library = libraries[index];
+
+      database[library.name] = {
+        ...database[library.name],
+        [library.version]: {
```

Comment:
how do you feel about only adding the separate version stats if it's more than 512 bytes different from the latest version? otherwise it'll all pretty much look the same to the user anyway :)

Reply:
given this any more thought? :)

I'm not sure what happened to my other comment about not including *any* size information for the *suggestions* other than the latest size. we'll never use the data about previous version suggestions in the audit.